src/valid/handles.rs

```diff
@@ -0,0 +1,337 @@
+    ) -> Result<(), InvalidHandleError> {
+        let &crate::Module {
+            ref constants,
+            ref entry_points,
+            ref functions,
+            ref global_variables,
+            ref types,
+        } = module;
+
+        // NOTE: Types being first is important. All other forms of validation depend on this.
+        for (this_handle, type_) in types.iter() {
+            let &crate::Type { name: _, ref inner } = type_;
+
+            match inner {
+                &crate::TypeInner::Scalar { .. }
+                | &crate::TypeInner::Vector { .. }
+                | &crate::TypeInner::Matrix { .. }
+                | &crate::TypeInner::ValuePointer { .. }
+                | &crate::TypeInner::Atomic { .. }
+                | &crate::TypeInner::Image { .. }
+                | &crate::TypeInner::Sampler { .. } => (),
+                &crate::TypeInner::Pointer { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Array { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+                &crate::TypeInner::Struct { ref members, .. } => {
+                    this_handle.check_dep_iter(members.iter().map(|m| m.ty))?;
+                }
+                &crate::TypeInner::BindingArray { base, .. } => {
+                    this_handle.check_dep(base)?;
+                }
+            }
+        }
+
+        let validate_type = |handle: Handle<crate::Type>| handle.check_valid_for_uniq(types);
+
+        for (this_handle, constant) in constants.iter() {
+            let &crate::Constant {
+                name: _,
+                specialization: _,
+                ref inner,
+            } = constant;
+            match *inner {
+                crate::ConstantInner::Scalar { .. } => (),
+                crate::ConstantInner::Composite { ty, ref components } => {
+                    validate_type(ty)?;
+                    this_handle.check_dep_iter(components.iter().copied())?;
+                }
+            }
+        }
+
+        let validate_constant = |handle: Handle<crate::Constant>| handle.check_valid_for(constants);
+
+        for (_handle, global_variable) in global_variables.iter() {
+            let &crate::GlobalVariable {
+                name: _,
+                space: _,
+                binding: _,
+                ty,
+                init,
+            } = global_variable;
+            validate_type(ty)?;
+            if let Some(init_expr) = init {
+                validate_constant(init_expr)?;
+            }
+        }
+
+        let validate_expressions = |expressions: &Arena<crate::Expression>,
+                                    local_variables: &Arena<crate::LocalVariable>|
+         -> Result<_, InvalidHandleError> {
+            for (this_handle, expression) in expressions.iter() {
+                match expression {
+                    &crate::Expression::Access { base, .. }
```

Comment:
We need to check the `index` of an `Access` expression as well.

Reply:
Oh, oof, good eye. Fixed with 4c5e2099.